packages/flutter/lib/src/widgets/framework.dart

```diff
@@ -1752,7 +1773,20 @@ class InheritedElement extends _ProxyElement {
       _inheritedWidgets = new Map<Type, InheritedElement>.from(incomingWidgets);
     else
       _inheritedWidgets = new Map<Type, InheritedElement>();
-    _inheritedWidgets[widget.runtimeType] = this;
+    final Type widgetType = widget.runtimeType;
+    if (_child != null && !_inheritedWidgets.containsKey(widgetType)) {
+      ElementVisitor visitor;
+      visitor = (Element child) {
+        if (child.runtimeType == widgetType)
+          return; // no point going further down this tree, they have one already
+        if (child._unsatisfiedDependencies != null &&
+            child._unsatisfiedDependencies.contains(widgetType))
+          child.dependenciesChanged(widgetType);
+        child.visitChildren(visitor);
```

Comment:
I suspect there's a better approach here.  _updateInheritance is called in a downward walk and you're generating a nested downward walk, which often will end up being n^2.  Instead, you should try passing information down the activate walk (e.g., in a function argument) that lets each element update its unsatisfied dependencies when we visit them during the normal course of the activate walk.  That will avoid the n^2.